web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx

```diff
@@ -896,6 +927,84 @@ export function WindowsDesktopAccessSpecSection({
   );
 }
 
+export function AdminRules({
+  value,
+  isProcessing,
+  validation,
+  onChange,
+}: SectionProps<RuleModel[], AdminRuleValidationResult[]>) {
+  function addRule() {
+    onChange?.([...value, newRuleModel()]);
+  }
+  function setRule(rule: RuleModel) {
+    onChange?.(value.map(r => (r.id === rule.id ? rule : r)));
+  }
+  function removeRule(id: string) {
+    onChange?.(value.filter(r => r.id !== id));
+  }
```

Comment:
> Additionally, if you wanted to wrap addRule in useCallback, it would have to have value as dep, so it wouldn't really work.

Why wouldn't it work?

Reply:
When memoizing a function that updates a state, we usually don't want to have that state in its deps, because that will cause the function to be re-created every time the same function updates it.
https://react.dev/reference/react/useCallback#updating-state-from-a-memoized-callback

So the correct way to say this is that it would work, but would be much less efficient.